core/src/main/java/org/elasticsearch/search/suggest/Suggest.java

```diff
@@ -369,6 +384,49 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             return builder;
         }
 
+        public static Suggestion<? extends Entry<? extends Option>> fromXContent(XContentParser parser) throws IOException {
+            ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser::getTokenLocation);
+            String typeAndName = parser.currentName();
+            // we need to extract the type prefix from the name and throw error if it is not present
+            int delimiterPos = typeAndName.indexOf(InternalAggregation.TYPED_KEYS_DELIMITER);
+            String type = null;
+            String name = null;
+            if (delimiterPos > 0) {
+                type = typeAndName.substring(0, delimiterPos);
+                name = typeAndName.substring(delimiterPos + 1);
+            } else {
+                throw new ParsingException(parser.getTokenLocation(),
+                        "Cannot parse suggestion response without type information. Set [" + RestSearchAction.TYPED_KEYS_PARAM
+                                + "] parameter on the request to ensure the type information is added to the response output");
+            }
+            Suggestion suggestion = null;
+            Function<XContentParser, Entry> entryParser = null;
+            /// the "size" parameter and the SortBy for TermSuggestion cannot be parsed from the response, use default values
+            switch (type) {
+            case Suggestion.NAME:
+                suggestion = new Suggestion(name, -1);
+                entryParser = Suggestion.Entry::fromXContent;
+                break;
+            case PhraseSuggestion.NAME:
+                suggestion = new PhraseSuggestion(name, -1);
+                entryParser = PhraseSuggestion.Entry::fromXContent;
+                break;
+            case TermSuggestion.NAME:
+                suggestion = new TermSuggestion(name, -1, SortBy.SCORE);
+                entryParser = TermSuggestion.Entry::fromXContent;
+                break;
+            case CompletionSuggestion.NAME:
+                suggestion = new CompletionSuggestion(name, -1);
+                entryParser = CompletionSuggestion.Entry::fromXContent;
+                break;
+            }
```

Comment:
Will add a TODO to re-evaluate this. I'm not sure this fits in 100% with NamedXContentRegistry since we not only need to look up a parser but also the right constructor for the suggestion. I will add this to the comment.